### PR TITLE
reinforce petsc options for the adjoint refs #25689

### PR DESCRIPTION
**Pull request by oanaoana**
WIP reinforcing the same solver options for both forward and adjoint problem refs #25689

### CODE REVIEW

**Comment by lynnmunday** [comment]
This fixes my problem, the petsc_option preconditoner is being applied to the adjoint problem.

**Comment by lynnmunday** [comment]
Later on we can implement something for each nonlinear system like modules/navier_stokes/src/executioners/SIMPLE.C 
 does for momentum/pressure/energy_petsc_options.  But this is good for now.  Thanks!

**Review by lindsayad — CHANGES_REQUESTED** [review]
We should add a test that verifies this is achieving what we want

**Comment by oanaoana** [comment]
Isn't `/modules/optimization/test/tests/optimizationreporter/nonlinear_material/main.i` sufficient? It does show that under `-ksp-view` the PC takes the input file specified values.

**Comment by lindsayad** [comment]
You could add `-ksp_view` and then in the `tests` spec add an `absent_out` parameter that ensures `ilu` is absent

**Comment by lindsayad** [comment]
I am curious why the options weren't being taken in the first place?

**Comment by oanaoana** [comment]
> I am curious why the options weren't being taken in the first place?

Clearly the solver options didn't trickle down to the adjoint in the subapp, but they were also never explicitly set. If the expectation is that they are automatically inherited it's tricky to identify where this fails and likely some refactoring to sort it out.

**Comment by lynnmunday** [comment]
> You could add `-ksp_view` and then in the `tests` spec add an `absent_out` parameter that ensures `ilu` is absent

I can make one that would fail if ilu is used.  Adjoint.c is a solverObject and not an executioner, I don't know if that would be affecting how it gets its petsc options.  It is weird it only happens in subapps.

File: modules/optimization/src/executioners/AdjointSolve.C [hunk 1]
```diff
@@ -52,6 +52,10 @@ AdjointSolve::solve()
   if (_inner_solve && !_inner_solve->solve())
     return false;
```
Comment (GiudGiud) [inline modules/optimization/src/executioners/AdjointSolve.C:54]: add a comment please

this shouldnt be necessary, but it is so let s remember why

Comment (oanaoana) [reply modules/optimization/src/executioners/AdjointSolve.C:54]: I don't understand; this isn't part of this PR.

Comment (GiudGiud) [reply modules/optimization/src/executioners/AdjointSolve.C:54]: the three lines below are?

Comment (lynnmunday) [reply modules/optimization/src/executioners/AdjointSolve.C:54]: This is what fixedPointSolve.C does when the inner_solve fails and then the return false is to catch when this solverObject fails in SteadyAndAdjoint.C

But this is not part of Oana PR.

Comment (GiudGiud) [reply modules/optimization/src/executioners/AdjointSolve.C:54]: I ll need to be more accurate with my clicks. One line too high and 3 people had to get involved

**Comment by lindsayad** [comment]
I don't care who adds it, but we still need a verification test

**Comment by lynnmunday** [comment]
I'll add a test.

**Comment by oanaoana** [comment]
> I don't care who adds it, but we still need a test

@lynnmunday  saved the day!

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
thanks for the fix!

**Pull request merged**